BadMessage interface

gregw opened the pull request:
Convert class BadMessageException to BadMessage interface that allows different types of BadMessage exceptions. This follows the pattern of the QuietException interface.

joakime commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTPTest.java:
```diff
@@ -206,7 +206,7 @@ public void testReceiveBadResponse(HttpCompliance compliance) throws Exception
         ExecutionException e = assertThrows(ExecutionException.class, () -> listener.get(5, TimeUnit.SECONDS));
         assertThat(e.getCause(), instanceOf(HttpResponseException.class));
-        assertThat(e.getCause().getCause(), instanceOf(BadMessageException.class));
+        assertThat(e.getCause().getCause(), instanceOf(BadMessage.RuntimeException.class));
```
Can't this just be `RuntimeBadMessageException` ?

gregw replied:
It could be just `BadMessage.class`, but it is a test, so no harm testing for a specific type.

joakime commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/ErrorPageTest.java:
```diff
@@ -390,11 +390,11 @@ public void testBadMessage() throws Exception
-            assertThat(response, Matchers.containsString("ERROR_EXCEPTION: org.eclipse.jetty.http.BadMessageException: 400: Unable to parse URI query"));
-            assertThat(response, Matchers.containsString("ERROR_EXCEPTION_TYPE: class org.eclipse.jetty.http.BadMessageException"));
+            assertThat(response, Matchers.containsString("ERROR_EXCEPTION: org.eclipse.jetty.http.BadMessage$RuntimeException: 400: Unable to parse URI query"));
+            assertThat(response, Matchers.containsString("ERROR_EXCEPTION_TYPE: class org.eclipse.jetty.http.BadMessage$RuntimeException"));
```
This bothers me, as inner classes for working with servlet error pages is confusing enough.
Can this please not be an inner class and be it's own top level?

gregw replied:
The Jetty-12 style is heavily using inner classes of interfaces.   It could be a top level class, but that is against the style we have been using.

gregw replied:
@joakime see QuietException, HttpFields, HttpCookie, HttpURI, Handler, Content, etc. as examples of this style.

joakime replied:
Great, but none of those examples are common for users to capture in a Servlet error page.

gregw replied:
@joakime  it is highly unlikely that webapps will map BadMessageException to an error page.   If they want an error page for 400 responses, they just map code 400 rather than a non standard exception.

But see latest changes anyway, as BadMEssageException is back as a top level class.

sbordet reviewed (changes requested):
I don't understand this change.

`BadMessage` is-an exception so should end with `Exception` in the name to follow the convention.

The 2 inner classes are never caught separately to act differently depending on the type, so I don't understand why we need 2 of them?

`BadMessage.IllegalArgumentException` is never thrown, only caught in 1 place where also `BadMessage.RuntimeException` is caught, so these changes seems to change nothing from having just one class?

gregw committed:
HttpMessage interface

Convert class BadMessageException to a HttpMessage.RuntimeException to allows different types of HttpExceptions exceptions.
This follows the pattern of the QuietException interface.

gregw commented:
I redid this is a less intrusive way (forced push sorry).

I've kept `BadMessageException`, but it is only used for 4xx responses. It extends the new `HttpException.RuntimeException` class.

All the handling of `BadMessageException` has been converted to checking for `HttpException`

gregw commented:
@sbordet just to record our hangout....
We can't have a tagging interface extend `Throwable` or `Exception` as they are not themselves interfaces.
So I can call it `HttpException` (even though `HttpThrowable` would be less inaccurate) like `QuietException` but is is-not-a `Exception` other than by convention.  I've added convenience static methods to do any casting back to a `Throwable` as needed.

gregw committed:
HttpMessage interface

Convert class BadMessageException to a HttpMessage.RuntimeException to allows different types of HttpExceptions exceptions.
This follows the pattern of the QuietException interface.

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/HttpReceiverOverHTTP.java:
```diff
@@ -550,7 +550,7 @@ public void badMessage(BadMessageException failure)
             HttpResponse response = exchange.getResponse();
             response.status(failure.getCode()).reason(failure.getReason());
-            failAndClose(new HttpResponseException("HTTP protocol violation: bad response on " + getHttpConnection(), response, failure));
+            failAndClose(new HttpResponseException("HTTP protocol violation: bad response on " + getHttpConnection(), response, HttpException.asThrowable(failure)));
```
Honestly, I would just cast: `_failure = (Throwable)failure;`.
I'd prefer to fail the cast, rather than having `_failure==null` in the unlikely case we mess up and make one non-exception class implement `HttpException`.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/BadMessageException.java:
```diff
@@ -16,56 +16,49 @@
+        assert code >= 400 && code < 500;
     }
 
+    @Override
     public int getCode()
     {
         return _code;
     }
 
+    @Override
     public String getReason()
     {
         return _reason;
```
No need for these, they are in the base class.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpException.java:
```diff
@@ -0,0 +1,128 @@
+
+    String getReason();
+
+    static Throwable asThrowable(HttpException httpException)
+    {
+        return httpException instanceof Throwable throwable ? throwable : null;
+    }
```
I would remove this and replace its usages with a plain cast to `Throwable`, which fails if we mess up.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpException.java:
```diff
@@ -0,0 +1,128 @@
+        return httpException instanceof Throwable throwable ? throwable : null;
+    }
+
+    static void rethrowRuntime(HttpException httpException)
```
I would prefer this to be `static RuntimeException asRuntimeException(HttpException httpException)` and in the places where it is used have `throw HttpException.asRuntimeException(httpException);`.

gregw replied:
Can't do that as we'd need a multi-variant return of Error or RuntimeException.
I'll rename to `throwAsUnchecked(HttpException)`

sbordet replied:
Not sure I follow. An `HttpException` cannot be an `Error`, and it is always a `RuntimeException`.

gregw replied:
@sbordet a HttpException is just a marker interface. Currently we apply it to `RuntimeException` and `IllegalArgumentException`, but in future we may tag an Error as a HttpException.  Unlikely, but having just made a whole PR about decoupling `HttpException` from the specific type of exception, it seams silly to bind it here.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpException.java:
```diff
@@ -0,0 +1,128 @@
+    @SuppressWarnings("serial")
+    class RuntimeException extends java.lang.RuntimeException implements HttpException
+    {
+        final int _code;
+        final String _reason;
```
Make them `private`.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpException.java:
```diff
@@ -0,0 +1,128 @@
+    @SuppressWarnings("serial")
+    class IllegalArgumentException extends java.lang.IllegalArgumentException implements HttpException
+    {
+        final int _code;
+        final String _reason;
```
Make them `private`.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java:
```diff
@@ -429,18 +429,17 @@ else if (status == HttpStatus.NO_CONTENT_204 || status == HttpStatus.NOT_MODIFIE
                 {
-                    throw new BadMessageException(INTERNAL_SERVER_ERROR_500, e.getMessage(), e);
+                    if (e instanceof HttpException httpException)
+                        HttpException.rethrowRuntime(httpException);
```
Like done elsewhere, you don't need this, you can just `throw e`.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpTester.java:
```diff
@@ -428,9 +428,9 @@ public boolean content(ByteBuffer ref)
+        public void badMessage(HttpException failure)
         {
-            throw failure;
+            HttpException.rethrowRuntime(failure);
```
See previous comment, I would prefer here: `throw HttpException.asRuntimeException(failure);`.
Or perhaps just the cast: `throw (RuntimeException)failure;` since all `HttpException`s are `RuntimeException`s.

gregw replied:
There is no reason why HttpExceptions must be RuntimeExceptions, they could be Errors in future.

sbordet replied:
No, that would be wrong, as `Error`s are only thrown by the JVM runtime, not applications.

gregw replied:
Which we might wish to catch and rethrow as a HttpException variant.   The whole point of this PR is decouple `HttpException` from any specific exception class.  A `throwAsUncheck(failure)` method is a reasonable way to keep this decouple, even if we never actually use that for `Error`s

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Dispatcher.java:
```diff
@@ -196,17 +196,17 @@ protected void forward(ServletRequest request, ServletResponse response, Dispatc
                     {
-                        // Only throw BME if not in Error Dispatch Mode
+                        // Only throw HttpException if not in Error Dispatch Mode
                         // This allows application ErrorPageErrorHandler to handle BME messages
```
```suggestion
                        // This allows application ErrorPageErrorHandler to handle the HttpException reason.
```

sbordet commented on jetty-ee9/jetty-ee9-nested/src/test/java/org/eclipse/jetty/ee9/nested/RequestTest.java:
```diff
@@ -220,7 +221,7 @@ public void testParamExtraction() throws Exception
                 return false;
             }
-            catch (BadMessageException e)
+            catch (Exception e)
```
```suggestion
            catch (Throwable e)
```

sbordet commented on jetty-ee9/jetty-ee9-nested/src/test/java/org/eclipse/jetty/ee9/nested/RequestTest.java:
```diff
@@ -575,7 +576,7 @@ public void testBadUtf8ParamExtraction() throws Exception
                 return false;
             }
-            catch (BadMessageException e)
+            catch (Exception e)
```
```suggestion
            catch (Throwable e)
```

sbordet commented on jetty-ee9/jetty-ee9-tests/jetty-ee9-test-client-transports/src/test/java/org/eclipse/jetty/ee9/test/client/transport/ServerTimeoutsTest.java:
```diff
@@ -325,9 +325,10 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
                     }
                 }
-                catch (BadMessageException x)
+                catch (Exception x)
```
```suggestion
                catch (Throwable x)
```

gregw committed:
BadMessage interface

Convert class BadMessageException to BadMessage interface that allows different types of BadMessage exceptions.
This follows the pattern of the QuietException interface.

gregw committed:
BadMessage interface

Convert class BadMessageException to BadMessage interface that allows different types of BadMessage exceptions.
This follows the pattern of the QuietException interface.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpException.java:
```diff
@@ -0,0 +1,123 @@
+ * <p>Exception sub-classes that implement this interface will be caught by the container
+ * and the {@link #getCode()} used to send a response.</p>
+ */
+@SuppressWarnings("serial")
```
What's with this suppressions? Just remove them, these exceptions are not serializable anyway and we don't want them to be.

gregw replied:
No idea. They were in the code to start with.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java:
```diff
@@ -429,18 +429,17 @@ else if (status == HttpStatus.NO_CONTENT_204 || status == HttpStatus.NOT_MODIFIE
-                    throw new BadMessageException(INTERNAL_SERVER_ERROR_500, e.getMessage(), e);
+                    if (e instanceof HttpException)
+                        throw e;
+
```
Remove newline.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpException.java:
```diff
@@ -0,0 +1,123 @@
+
+    static void throwAsUnchecked(HttpException httpException)
+    {
+        if (httpException instanceof Throwable throwable)
+            ExceptionUtil.ifExceptionThrowUnchecked(throwable);
```
I'd rather fail a cast than silently ignore an implementation mistake here.
So perhaps just `ExceptionUtil.ifExceptionThrowUnchecked((Throwable)httpException);`, although I would prefer this method to return the exception, and the caller to throw (in this way the method can be reused also to store the returned exception in a field, etc.).

gregw replied:
ExceptionUtil definitely can't return the exception, as it may be an Error.  Reusing that method also makes this best as a throwing method.

gregw committed:
BadMessage interface

Convert class BadMessageException to BadMessage interface that allows different types of BadMessage exceptions.
This follows the pattern of the QuietException interface.

gregw commented:
I've no idea why this branch suffers from so many flaky tests.  But I can see nothing in the CI that is related to this change.

olamy commented:
` testDownloadWithCloseBeforeContent(Transport)[1] – org.eclipse.jetty.test.client.transport.HttpClientStreamTest` is already flaky in jetty-12.0.x for many builds https://jenkins.webtide.net/job/jetty.project/job/jetty-12.0.x/1092/

olamy commented:
issue created https://github.com/eclipse/jetty.project/issues/9485 
best to mark it as flaky using `@Tag("flaky")`
